RFC: Add register_run_asset event

clairelin135 opened the pull request:
Adds a new Dagster event type `REGISTER_RUN_ASSET`.

This event is yielded after an asset job run begins for every single asset that is selected in the job. The goal of this feature is to enable querying for runs that intended to generate certain assets (e.g. job that intends to materialize 3 assets fails during materialization of second asset).

This will enable warnings in Dagit on the asset details page and the asset graph nodes. In the future, we can enable this for partitioned assets to generate views of all runs across a certain asset's partition. If this event becomes too noisy in run logs, we can consider ways to condense the logs.

clairelin135 commented on python_modules/dagster/dagster/core/events/__init__.py:
```diff
@@ -186,6 +187,8 @@ class DagsterEventType(Enum):
 PIPELINE_RUN_STATUS_TO_EVENT_TYPE = {v: k for k, v in EVENT_TYPE_TO_PIPELINE_RUN_STATUS.items()}
 
+# TODO: Add ASSET_INTENT_TO_MATERIALIZE event to access assets in run
+# when error occurs before materialization or observation occurs
```
Plan to handle this in next PR

clairelin135 commented on python_modules/dagster/dagster/core/definitions/pipeline_definition.py:
```diff
@@ -412,6 +412,29 @@ def solids(self):
+            if node_name in step_keys:
+                asset_keys_by_step[node_name] = []
+                for output_def in node_def.output_dict.values():
+                    asset_keys_by_step[node_name].append(output_def.hardcoded_asset_key)
```
@OwenKephart this assumes that all asset keys are strings. I saw a note that we're no longer supporting dynamic asset keys in 0.15.0 so I lean toward not supporting this feature on dynamic asset keys, what do you think?

sryza commented:
+1 to both things @prha said

prha commented:
nit: 'intent_to_materialize` seems a little too on the nose to me.  what bout something like:
* register_run_asset
* register_asset_plan
* plan_asset_run
* asset_run_created 
* etc.

clairelin135 commented:
@prha I like `register_run_asset`! Going to switch this over

prha commented on python_modules/dagster/dagster/core/instance/__init__.py:
```diff
@@ -1130,17 +1157,30 @@ def logs_after(
-        )
+        from dagster.core.events import DagsterEventType
+
+        # We remove REGISTER_RUN_ASSET events here because these events
```
This feels a little low-level to do the filtering here.  What do you think about letting these pass through the instance / storage APIs, and just filtering either in the graphql or in the JS layer?

prha commented:
cc @alangenfeld 

I think this is looking good to me... I'll feel better about landing this after we have some stacked changes of what the read side will look like though

clairelin135 commented:
@prha got it, thanks! I'll get started on the read-side for these events and we can discuss after I put diffs out.

alangenfeld commented:
Why N events instead of a single event?

alangenfeld commented on python_modules/dagster/dagster/core/execution/plan/outputs.py:
```diff
@@ -22,6 +23,7 @@ class StepOutputProperties(
             ("is_dynamic", bool),
             ("is_asset", bool),
             ("should_materialize", bool),
+            ("asset_key", Optional[AssetKey]),
```
its a pain to backcompat but should asset_key subsume / deprecate is_asset? Its information duplication right or am i missing something?

clairelin135 replied:
In most cases the information is the same, but for dynamic asset keys `is_asset` will return `True`, but `asset_key` will only return a value for hardcoded asset keys. We plan on deprecating dynamic asset keys by 0.15.0

clairelin135 replied:
I think that when we fully deprecate dynamic asset keys we can deprecate `is_asset` and just use `asset_key`.

alangenfeld commented on python_modules/dagster/dagster/core/instance/__init__.py:
```diff
@@ -932,6 +932,29 @@ def _ensure_persisted_execution_plan_snapshot(
+                for output in step.outputs:
+                    asset_key = output.properties.asset_key
+                    if asset_key:
+                        event = DagsterEvent.register_run_asset(pipeline_name, asset_key)
+                        event_record = EventLogEntry(
+                            user_message="",
+                            level=logging.INFO,
+                            pipeline_name=pipeline_name,
+                            run_id=pipeline_run.run_id,
+                            error_info=None,
+                            timestamp=time.time(),
+                            dagster_event=event,
+                        )
+                        self.handle_new_event(event_record)
```
this diverges from the pattern normal to the static constructors on `DagsterEvent` - i assume to avoid logging all the output? 

If we had 1 instead of N events it seems we could avoid going out of our way to hide this information which users may find valuable in some circumstances

alangenfeld replied:
what if we just dropped he log level to `debug` for "hiding" since the default dagit experience is to hide debug logs (i believe)? This does seem like useful debug information.

clairelin135 replied:
Yep, I think this will be useful debug information, I'll switch the log level to `debug` instead. The pattern diverges from the static constructors on `DagsterEvent` because no `pipeline_context` or `step_context` object has been created.

clairelin135 replied:
Hm...seems like we display all `DagsterEvent`s by default in Dagit and only filter debug logs that are not `DagsterEvent`s. I could add special casing though on the front end to hide events of this type

alangenfeld commented:
"register_run_asset" feels odd to me personally. Most of our existing events are `<noun>_<state>` so something like `asset_materializations_planned` feels better to me.

sryza commented:
> Why N events instead of a single event?

@alangenfeld so we can query by asset key.

alangenfeld commented on python_modules/dagster/dagster/core/storage/event_log/sqlite/sqlite_event_log.py:
```diff
@@ -232,11 +232,6 @@ def store_event(self, event):
             conn.execute(insert_event_statement)
 
         if event.is_dagster_event and event.dagster_event.asset_key:
-            check.invariant(
-                event.dagster_event_type == DagsterEventType.ASSET_MATERIALIZATION
-                or event.dagster_event_type == DagsterEventType.ASSET_OBSERVATION,
-                "Can only store asset materializations and observations in index database",
-            )
```
why remove instead of update?

clairelin135 replied:
no strong reason here...I didn't feel like the check was necessary, but I've readded and updated the check now

alangenfeld reviewed (commented):
ah duh I somehow forgot about the `asset_key` column on the event logs 

re: names, if we want to be terse `asset_planned` i think is acceptable. I personally like the `planned` suffix since these are derived from the `execution plan`

clairelin135 commented:
@alangenfeld hey alex, appreciate the detailed feedback! I'm going to switch the event name to `asset_materialization_planned`

clairelin135 commented:
@alangenfeld I've changed the event name to `asset_materialization_planned` and updated the logs to display in Dagit when debug is selected. I think this is ready for you to take another look!

alangenfeld commented on python_modules/dagster/dagster/core/events/__init__.py:
```diff
@@ -190,6 +192,7 @@ class DagsterEventType(Enum):
 ASSET_EVENTS = {
     DagsterEventType.ASSET_MATERIALIZATION,
```
stale?

alangenfeld commented on python_modules/dagster/dagster/core/instance/__init__.py:
```diff
@@ -932,6 +932,29 @@ def _ensure_persisted_execution_plan_snapshot(
+                for output in step.outputs:
+                    asset_key = output.properties.asset_key
+                    if asset_key:
+                        event = DagsterEvent.asset_materialization_planned(pipeline_name, asset_key)
+                        event_record = EventLogEntry(
+                            user_message="",
+                            level=logging.DEBUG,
+                            pipeline_name=pipeline_name,
+                            run_id=pipeline_run.run_id,
+                            error_info=None,
+                            timestamp=time.time(),
+                            dagster_event=event,
+                        )
+                        self.handle_new_event(event_record)
```
thoughts on putting this in line with the other DagsterEvent static constructors and logging there? Its a weird pattern but I think there is value in consistency.

clairelin135 replied:
Sure--I've updated the code now to log the dagster event in a static constructor.

alangenfeld commented on python_modules/dagster/dagster/core/storage/event_log/sqlite/sqlite_event_log.py:
```diff
@@ -230,9 +230,11 @@ def store_event(self, event):
+                "Can only store asset materializations, materialization_planned, and observations in index database",
             )
+
             # mirror the event in the cross-run index database
             with self.index_connection() as conn:
                 conn.execute(insert_event_statement)
```
<!-- GRAPHITE_METADATA {"trueLocation":{"startLine":235,"endLine":239}} -->
<i class='graphite__hidden'>[Re: lines 235 to 239]<br/><br/></i>

why remove instead of update?

clairelin135 replied:
No strong reason here--I removed it earlier because I didn't feel that the check was necessary. I've updated it now though to contain the new event type

alangenfeld commented on python_modules/dagster/dagster/core/instance/__init__.py:
```diff
@@ -932,6 +932,29 @@ def _ensure_persisted_execution_plan_snapshot(
+                        event = DagsterEvent.asset_materialization_planned(pipeline_name, asset_key)
+                        event_record = EventLogEntry(
+                            user_message="",
+                            level=logging.DEBUG,
```
might be good to leave reasoning next to the debug log level setting

clairelin135 replied:
Okay, added a comment next to the setting

alangenfeld commented on python_modules/dagster/dagster/core/events/__init__.py:
```diff
@@ -259,6 +266,16 @@ def log_resource_event(log_manager: DagsterLogManager, event: "DagsterEvent") ->
     log_manager.log_dagster_event(level=log_level, msg=event.message or "", dagster_event=event)
 
 
+def log_asset_materialization_planned_event(
+    log_manager: DagsterLogManager, event: "DagsterEvent"
+) -> None:
+    # asset_materialization_planned events have a log level "DEBUG" in order to hide these
+    # events by default in Dagit. Modifying filtering to select DEBUG events will show these events
+    # in Dagit run logs.
+    log_level = logging.DEBUG
+    log_manager.log_dagster_event(level=log_level, msg=event.message or "", dagster_event=event)
```
no need to cargo-cult the pattern here and refactor when single use  - this can just be inlined

clairelin135 replied:
✅

alangenfeld commented on python_modules/dagster/dagster/core/instance/__init__.py:
```diff
@@ -970,6 +987,10 @@ def create_run(
             pipeline_code_origin=pipeline_code_origin,
         )
+
+        if execution_plan_snapshot:
+            self._log_asset_materialization_planned_events(pipeline_run, execution_plan_snapshot)
+
         return self._run_storage.add_run(pipeline_run)
```
might be good to sequence the event writes after the run gets added to the db? incase the run add fails

clairelin135 replied:
✅

alangenfeld reviewed (commented):
@prha what are your thoughts on this debug log level hide the event thing

clairelin135 commented:
@alangenfeld I don't feel strongly either way about whether to make these events viewable in Dagit or not, though I don't think users will find them very useful (the run will already show which assets intend to be materialized, and these events are only responsible for populating asset run information in Dagit). Defer to you to make the final call though

alangenfeld reviewed (approved):
> Defer to you to make the final call though

I don't think I'm the best person to make final call on asset product experience questions. I don't have broader context, so things like

> the run will already show which assets intend to be materialized

I don't know how that currently is presented. 

As prha pointed out, this last bit is very easy to change so I will accept the PR and you can land it with the behavior you feel best about.

alangenfeld commented on python_modules/dagster-graphql/dagster_graphql/schema/pipelines/pipeline.py:
```diff
@@ -358,6 +359,11 @@ def resolve_assets(self, graphene_info):
+        events = [
+            event
+            for event in events
+            if event.dagster_event_type != DagsterEventType.ASSET_MATERIALIZATION_PLANNED
```
[1]

alangenfeld commented on python_modules/dagster-graphql/dagster_graphql/implementation/execution/__init__.py:
```diff
@@ -151,6 +151,11 @@ def _get_error_observable(observer):
+        events = [
+            event
+            for event in events
+            if event.dagster_event_type != DagsterEventType.ASSET_MATERIALIZATION_PLANNED
```
[1]

prha replied:
I forgot about the offset cursors...  I guess easiest thing to do is to just filter these events client-side?

alangenfeld replied:
ya, that or fix pagination to not be offset based 🙂

alangenfeld reviewed (commented):
[1] I believe filtering here is going to cause problems with the offset based pagination we currently have

clairelin135 commented:
@alangenfeld got it. I can adjust the filtering to filter client-side instead of here
